npms/skale-observer/observer.mjs

```diff
@@ -0,0 +1,1078 @@
+    if( opts && opts.details ) {
+        if( log.verboseGet() >= log.verboseReversed().trace ) {
+            opts.details.write( cc.debug( "Have " ) + cc.info( cntSChains ) +
+                cc.debug( " S-Chain(s) to load..." ) + "\n" );
+        }
+    }
+    const arrSChains = [];
+    for( let idxSChain = 0; idxSChain < cntSChains; ++ idxSChain ) {
+        if( opts && opts.bStopNeeded )
+            break;
+        const joSChain = await loadSChain( addressFrom, idxSChain, null, cntSChains, opts );
+        if( ! joSChain )
+            break;
+        arrSChains.push( joSChain );
+    }
+    if( opts && opts.details ) {
+        if( log.verboseGet() >= log.verboseReversed().trace ) {
+            opts.details.write( cc.success( "All " ) + cc.info( cntSChains ) +
+                cc.debug( " S-Chain(s) loaded:" ) + cc.j( arrSChains ) + "\n" );
+        }
+    }
+    return arrSChains;
+}
+
+export async function loadCachedSChainsSimplified( addressFrom, opts ) {
+    if( ! opts.imaState )
+        throw new Error( "Cannot load S-Chains in observer, no imaState is provided" );
+    if( opts && opts.details ) {
+        if( log.verboseGet() >= log.verboseReversed().trace )
+            opts.details.write( cc.debug( "Will request all S-Chain(s) hashes..." ) + "\n" );
+    }
+    const arrSChainHashes =
+        await opts.imaState.joSChainsInternal.callStatic.getSchains(
+            { from: addressFrom } );
+    const cntSChains = arrSChainHashes.length;
+    if( opts && opts.details ) {
+        if( log.verboseGet() >= log.verboseReversed().trace ) {
+            opts.details.write( cc.debug( "Have all " ) + cc.info( cntSChains ) +
+                cc.debug( " S-Chain(s) hashes: " ) + cc.j( arrSChainHashes ) + "\n" );
+        }
+    }
+    const arrSChains = [];
+    for( let idxSChain = 0; idxSChain < cntSChains; ++ idxSChain ) {
+        if( opts && opts.bStopNeeded )
+            break;
+        const strSChainHash = arrSChainHashes[idxSChain];
+        const strSChainName =
+            await opts.imaState.joSChainsInternal.callStatic.getSchainName(
+                strSChainHash, { from: addressFrom } );
+        if( opts && opts.details ) {
+            if( log.verboseGet() >= log.verboseReversed().trace ) {
+                opts.details.write( cc.debug( "S-Chain " ) + cc.notice( idxSChain ) +
+                    cc.debug( " hash " ) + cc.notice( strSChainHash ) +
+                    cc.debug( " corresponds to S-Chain name " ) + cc.notice( strSChainName ) +
+                    "\n" );
+            }
+        }
+        if( opts && opts.bStopNeeded )
+            break;
+        const joSChain =
+            await loadSChain(
+                addressFrom, idxSChain, strSChainHash, cntSChains, opts
+            );
+        if( ! joSChain )
+            break;
+        arrSChains.push( joSChain );
+    }
+    return arrSChains;
+}
+
+export async function loadSChainsConnectedOnly(
```

Comment:
Here as well, you can do just couple multicalls:
You can load all schains and then check are they connected by multicall
Right now function is doing 21 calls on mainnet, but with multicall you can decrease the number to 2 calls:
1. getSchains
2. multicall(getSchainName)

Reply:
Done